### PR TITLE
React Issue 2: Submit Feedback to backend

### PR DESCRIPTION
**Pull request by chen-daniel**
Closes React #2

### CODE REVIEW

**Comment by chen-daniel** [comment]
[React Issue 2: Submit Feedback to backend](https://trello.com/c/5Ajoi977/71-react-issue-2-submit-feedback-to-backend)

File: test/models/feedback_test.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+class FeedbackTest < ActionDispatch::IntegrationTest
+  test 'should create a Feedback with a name and comments' do
+    assert_difference 'Feedback.count', 1 do
+      fb = Feedback.create!(name: 'Daniel', comments: 'Feedfront')
+      assert_equal fb, Feedback.last
+    end
+  end
+
+  test 'should reject a Feedback that is missing a name' do
+    assert_no_difference 'Feedback.count' do
```
Comment (raj-al-ghul) [inline test/models/feedback_test.rb]: is there something else we can assert on in the invalid case?

File: app/controllers/api/feedbacks_controller.rb [hunk 2]
```diff
@@ -1,7 +1,20 @@
 module Api
   class FeedbacksController < ApplicationController
     def create
-      # Implement your create action
+      fb = Feedback.new(feedback_params)
+
+      if !fb.valid?
```
Comment (raj-al-ghul) [inline app/controllers/api/feedbacks_controller.rb]: personally, unless its a guard clause, i like the happy case first
(no need to change)

File: app/javascript/image_sharing/components/FeedbackForm.js [hunk 3]
```diff
@@ -1,21 +1,102 @@
...
+          this.feedbackFormStatus.alert = err.message;
+        }
+      });
+  };
+
+  render() {
+    return (
+      <Container>
+        <Alert
+          isOpen={Boolean(this.feedbackFormStatus.success)}
```
Comment (raj-al-ghul) [inline app/javascript/image_sharing/components/FeedbackForm.js]: `Boolean` is good, but in general I prefer to use `!!` because accidentally doing `new Boolean` can be bad https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Boolean#Description

File: app/javascript/image_sharing/components/FeedbackForm.js [hunk 4]
```diff
@@ -1,21 +1,102 @@
...
+          toggle={() => this.handleDismiss('success')}
+          color="info"
+        >
+          {this.feedbackFormStatus.success}
+        </Alert>
+        <Alert
+          isOpen={Boolean(this.feedbackFormStatus.alert)}
```
Comment (raj-al-ghul) [inline app/javascript/image_sharing/components/FeedbackForm.js]: let's refactor the flash messaging to not be like the rails version

File: app/javascript/image_sharing/components/FeedbackForm.js [hunk 5]
```diff
@@ -1,21 +1,102 @@
...
+    this.feedback[e.target.name] = e.target.value;
+    this.feedbackFormStatus[e.target.name] = '';
+  }
+
+  handleDismiss = (e) => {
+    this.feedbackFormStatus[e] = '';
+  }
+
+  submitForm = (e) => {
+    e.preventDefault();
```
Comment (raj-al-ghul) [inline app/javascript/image_sharing/components/FeedbackForm.js]: why is this needed if we're using `Form`'s `onSubmit`

File: app/javascript/image_sharing/components/FeedbackForm.js [hunk 6]
```diff
@@ -1,21 +1,102 @@
...
+          {this.feedbackFormStatus.alert}
+        </Alert>
+        <Form onSubmit={this.submitForm}>
+          <FormGroup>
+            <Label for="feedbackName">Your Name</Label>
+            <Input
+              type="text"
+              name="name"
+              id="feedbackName"
+              onChange={this.handleChange}
```
Comment (raj-al-ghul) [inline app/javascript/image_sharing/components/FeedbackForm.js]: lets do a change per value, simpler & easier to read > slightly less code

File: test/models/feedback_test.rb [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+      assert_raises(ActiveRecord::RecordInvalid) do
```
Comment (raj-al-ghul) [inline test/models/feedback_test.rb]: can we check the validation errors on the model, and consolidate tests

File: app/controllers/api/feedbacks_controller.rb [hunk 8]
```diff
@@ -1,7 +1,20 @@
...
+        render status: :bad_request, json: fb.errors.messages
```
Comment (raj-al-ghul) [inline app/controllers/api/feedbacks_controller.rb]: `422 Unprocessable Entity` response status code indicates that the server understands the content type of the request entity, and the syntax of the request entity is correct, but it was unable to process the contained instructions

vs. 

`400 Bad Request` response status code indicates that the server cannot or will not process the request due to something that is perceived to be a client error (e.g., malformed request syntax, invalid request message framing, or deceptive request routing).

File: test/controllers/api/feedbacks_controller_test.rb [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    assert_no_difference 'Feedback.count' do
+      post api_feedbacks_path, params: { feedback: { name: 'Nice' } }
+    end
+    assert_response :bad_request
+    res = JSON.parse(response.body)
+    assert_equal res['comments'].length, 1
+    assert_equal res['comments'][0], "can't be blank"
+  end
+
+  test 'create should fail to create feedback with incorrect params' do
```
Comment (raj-al-ghul) [inline test/controllers/api/feedbacks_controller_test.rb]: isn't really a necessary test

File: test/controllers/api/feedbacks_controller_test.rb [hunk 10]
```diff
@@ -0,0 +1,42 @@
+require 'test_helper'
+
+class FeedbacksControllerTest < ActionDispatch::IntegrationTest
+  test 'create should successfully create a feedback and respond with ok' do
+    assert_difference 'Feedback.count', 1 do
+      post api_feedbacks_path, params: { feedback: { name: 'Daniel', comments: 'Nice' } }
+    end
```
Comment (raj-al-ghul) [inline test/controllers/api/feedbacks_controller_test.rb]: should check if the created record has the data that we expect

File: app/javascript/image_sharing/test/components/FeedbackForm.spec.js [hunk 11]
```diff
@@ -23,4 +25,203 @@ describe('<FeedbackForm />', () => {
...
+      assert(e.preventDefault.calledOnce);
+      assert(stub.calledOnceWithExactly('/api/feedbacks', { name: '', comments: '' }));
+      assert.strictEqual(wrapper.instance().nameFeedback, "can't be blank");
+      assert.strictEqual(wrapper.instance().commentsFeedback, "can't be blank");
+      assert.strictEqual(wrapper.instance().alertMessage, undefined);
+      stub.restore();
+    });
+  });
+
+  it('should fail to submit feedback with empty comments', () => {
```
Comment (raj-al-ghul) [inline app/javascript/image_sharing/test/components/FeedbackForm.spec.js]: (while normally we'd do client side validations as we discussed)
the distinction of this case isn't necessary because in this case our client isn't handling the errors.
we just need one case where we test that our client handles the reject in a way we expect.

The reject in the test is handled by our `sinon.stub()` so it will reject regardless of the data we setup

File: app/javascript/image_sharing/test/components/FeedbackForm.spec.js [hunk 12]
```diff
@@ -23,4 +25,203 @@ describe('<FeedbackForm />', () => {
...
+  });
+
+  it('adding an error alert message and color should show a danger alert in the UI', () => {
+    const wrapper = mount(<FeedbackForm />);
+
+    wrapper.instance().alertMessage = 'This is a serious alert';
+    wrapper.instance().alertColor = 'danger';
+    wrapper.update();
+
+    const alert = wrapper.find({ isOpen: true, color: 'danger' });
```
Comment (raj-al-ghul) [inline app/javascript/image_sharing/test/components/FeedbackForm.spec.js]: imo more readable if we find an `Alert` and then assert on its props
also not clear where `isOpen` is coming from (may be an artifact of using `mount` in this case vs. `shallow`)

File: app/javascript/image_sharing/test/components/FeedbackForm.spec.js [hunk 13]
```diff
@@ -23,4 +25,203 @@ describe('<FeedbackForm />', () => {
...
+    assert.strictEqual(wrapper.instance().commentsFeedback, undefined);
+  });
+
+  it('handleDismiss should reset corresponding feedbackFormStatus to an empty string', () => {
+    const wrapper = mount(<FeedbackForm />);
+    wrapper.instance().alertMessage = 'This is an alert';
+    wrapper.instance().alertColor = 'danger';
+
+    wrapper.instance().handleDismiss();
+    assert.strictEqual(wrapper.instance().alertMessage, undefined);
```
Comment (raj-al-ghul) [inline app/javascript/image_sharing/test/components/FeedbackForm.spec.js]: do we have a test for the conditional rendering of `Alert`?

File: app/javascript/image_sharing/test/components/FeedbackForm.spec.js [hunk 14]
```diff
@@ -23,4 +25,203 @@ describe('<FeedbackForm />', () => {
...
+    wrapper.instance().comments = 'May the Force Push be with you';
+    wrapper.update();
+    nameInput = wrapper.find('#feedbackName').first();
+    commentInput = wrapper.find('#feedbackComments').first();
+
+    assert.strictEqual(nameInput.prop('value'), 'Luke');
+    assert.strictEqual(commentInput.prop('value'), 'May the Force Push be with you');
+  });
+
+  it('setters should update corresponding feedback and feedbackFormStatus field', () => {
```
Comment (raj-al-ghul) [inline app/javascript/image_sharing/test/components/FeedbackForm.spec.js]: update name?

also is this already tested in the above tests?

Comment (chen-daniel) [reply app/javascript/image_sharing/test/components/FeedbackForm.spec.js]: This test is testing the setters whereas the above is testing the UI. Should I combine them?

File: test/controllers/api/feedbacks_controller_test.rb [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+    assert_equal Feedback.last[:name], 'Daniel'
```
Comment (raj-al-ghul) [inline test/controllers/api/feedbacks_controller_test.rb]: since this isn't a hash, you can do `Feedback.last.name`

File: app/javascript/image_sharing/test/components/FeedbackForm.spec.js (hunk 14 above)
Comment (raj-al-ghul) [reply app/javascript/image_sharing/test/components/FeedbackForm.spec.js]: we talked about using `simulate` instead of having two separate tests

File: app/javascript/image_sharing/components/FeedbackForm.js [hunk 16]
```diff
@@ -1,21 +1,112 @@
...
-    </Container>
-  );
+import { Container, Form, FormGroup, Label, Input, Button, FormFeedback, Alert } from 'reactstrap';
+import { observer } from 'mobx-react';
+import { observable, toJS, action } from 'mobx';
+import * as api from '../utils/helper';
+
+@observer
+export default class FeedbackForm extends React.Component {
+  @observable feedbackFormStatus = {
```
Comment (raj-al-ghul) [inline app/javascript/image_sharing/components/FeedbackForm.js]: i dont think this is used anywhere anymore

File: app/javascript/image_sharing/components/FeedbackForm.js [hunk 17]
```diff
@@ -1,21 +1,107 @@
...
+
+  @observable alertMessage = undefined;
+  @observable alertColor = undefined;
+
+  @action
+  handleDismiss = () => {
+    this.alertMessage = undefined;
+  }
+
+  submitForm = (e) => {
```
Comment (raj-al-ghul) [inline app/javascript/image_sharing/components/FeedbackForm.js]: since we're editing `observable`s in this method lets make this an `@action` as well
